e2eAIOK/ModelAdapter/training/train.py

```diff
@@ -112,11 +112,18 @@ def _dist_wrapper(self):
         """            
         if ext_dist.my_size > 1:
             self.logger.info("training with DistributedDataParallel")
-            self.model = ext_dist.DDP(self.model)
-            self.model.loss = MethodType(lambda obj,*args: obj.module.loss(*args) ,self.model) #  obj stands for model, both original model and transferrable model
-            if self.is_transferrable: # dispatch method to model.module
-                self.model.get_backbone = MethodType(lambda obj: obj.module.get_backbone() ,self.model) # obj stands for model
-                self.model.get_training_metrics = MethodType(lambda obj,*args: obj.module.get_training_metrics(*args) ,self.model) #  obj stands for model
+            if self.is_transferrable:
+                self.model.backbone = ext_dist.DDP(self.model.backbone)
+                self.model.backbone.loss = MethodType(lambda obj,*args: obj.module.loss(*args) ,self.model.backbone) #  obj stands for model, both original model and transferrable model
+                if self.model.distiller is not None and not self.model.distiller.is_frozen:
+                    self.model.distiller = ext_dist.DDP(self.model.distiller) # if distiller require_grad=False, may fail in this step, just skip
+                if self.model.adapter is not None:
+                    self.mode.adapter = ext_dist.DDP(self.mode.adapter)
```

Comment:
what will happen if the configuration not either met line 118 condition or line 120 condition? Does that mean it will directly use no DDP?

Reply:
118 and 120 is to warp the submodule of transferrable mode.
Here is the full code, if it's not a transferrable model, it will use normal DDP. 
![image](https://user-images.githubusercontent.com/82487983/220854059-e622ae3f-e26a-443d-bc74-df9ce06926f9.png)

Reply:
which line is for normal DPP? if configuration shows it is 'is_transferrable' but not fall into line 118 and line 120, it also won't fall into else. So it will leave ext_dist.my_size > 1 condition.

Reply:
The components of transferrable model are: backbone, finetuner, distiller, adapter.
line 116 deals with backbong.
Finetuner's function is to apply parmeter, no need for training.
line 118 and 120 will deal distiller and adapter.
No other component will appear in transferrable model.

Reply:
will there be any case falls to self.model.distiller is not None and self.model.distiller.is_frozen ?
Accoding to SDL guideline, usually we should put some else raise to cover any unexpected scenario.
I felt currently you only covered expected scenarios, but will there be any user configuration that goes to condition not covered above?

Reply:
"self.model.distiller is not None and self.model.distiller.is_frozen" In this case, as it's parameter is frozen, distiller will not be trained and don't need to apply DDP.
"usually we should put some else raise to cover any unexpected scenario." - Yes, but in this case, I don't find any unexpected scenarios, welcome for any other thoughts~